### ISSUE TITLE
Anchor strides with MobileNet

### ISSUE THREAD

**Issue by Cre1th** [fizyr/keras-retinanet#342]
I observed that the separable convolutions with stride 2 in MobileNet round up not down (like ResNet) if the image dimension is not divisible by 2. Therefore a 240x**135** score map is transformed to 120x**68**. How do the pixels of the new score maps correspond to the coordinates in the original image? Does this in any way affect the placement of the anchors?

**Comment by lvaleriu** [comment]
There is actually a pull request (https://github.com/fizyr/keras-retinanet/pull/298) pending that will solve this.

**Comment by Cre1th** [comment]
This pull request takes care of the problem how many anchors are generated. That is good but I am concerned how the stride is chosen: Take for example an image of 135x135. Separable conv with stride 2 gives you 68x68. Now is the stride used for the shifting of the anchors 2 (stride of operation) or 135/68=1.985 (stride of the feature map in regard to the original image). For early layers this does not make a huge difference but for very deep layers it does.

**Comment by hgaiser** [comment]
I see your point but the strides are always assumed to be a power of 2. This is as far as I know how the original work also implements it, but it might be interesting to perform some comparison between the two methods if you feel like it.

**Comment by Cre1th** [comment]
Have you trained RetinaNet using MobileNet Backbone? I am asking because I am working on a similar network (R-FCN) and I experience the problem, that my boxes are shifted slightly to the bottom right when switching from ResNet50 to MobileNet - even though I changed nothing but backbone. This is true for my region proposal network as well (which is similar to the direct approach RetinaNet uses)

**Comment by hgaiser** [comment]
> Have you trained RetinaNet using MobileNet Backbone?

I haven't, nope.



> I am asking because I am working on a similar network (R-FCN) and I experience a problem, that my boxes are shifted slightly to the bottom right when switching from ResNet50 to MobileNet - even though I changed nothing but the backbone. This is true for my region proposal network as well (which is similar to the direct approach RetinaNet uses).
> 
> It would be very helpful to know if other people have the same issue when training the network.

Perhaps @lvaleriu has more information for you on how MobileNet compares to ResNet. I haven't worked with MobileNet yet (but do intend to).

> A short sidenote to #340: The network can still be trained even with the resnet preprocessing. It should only take a little longer as the weights have to be adjusted to fit the new input style.

Yeah I suppose you're right, but as you said training will take a lot longer.

**Comment by lvaleriu** [comment]
I can only confirm that training takes much more time. At this moment I never achieved a result close enough (to be useful) to the one obtained when training retinanet with resnet on coco.
My main issue is the training speed... whenever this will be fixed I’ll try to modify the processing of the input.

I’m actually in vacation till the end of month so I won’t be very active.

**Comment by Cre1th** [comment]
Have you taken a look at the actual predictions in order to check if they align with the ground truth or just the coco metric results?

**Comment by lvaleriu** [comment]
Nope. I only relied on the coco evaluation metric. I will try that soon!

**Comment by jonla1** [comment]
I did another training and the error is no longer present, so the anchors seem to be correct. I therefore will close this isse.
